### ISSUE TITLE
Too many bugs like memory leaks or cpu usage too high.

### ISSUE THREAD

**Issue by MrLiuYunPing** [aws/aws-sdk-ios#968]
### To help us solve your problem better, please answer the following list of questions.

* What service are you using?
Iot
* In what version of SDK are you facing the problem?
2.6.21
* Is the issue limited to Simulators / Actual Devices?
NO
* Can your problem be resolved if you bump to a higher version of SDK?
NO
* Is this problem related to specific iOS version?
NO
* How are you consuming the SDK? CocoaPods / Carthage / Prebuilt frameworks?
CocoaPods
* Can you give us steps to reproduce with a minimal, complete, and verifiable example? Please include any specific network conditions that might be required to reproduce the problem.
Too many bugs like memory leaks or cpu usage too high.
Memory leaks in runtime when reading .p12 method and other methods.
When network is lost,mqtt reconnection will make cpu usage too high in about 3 second time.

If you need help with understanding how to implement something in particular then we suggest that you first look into our [developer guide](https://docs.aws.amazon.com/aws-mobile/latest/developerguide/aws-mobile-android-and-iOS.html ). You can also simplify your process of creating an application by using [Mobile Hub](https://console.aws.amazon.com/mobilehub/home#/).

**Comment by cbommas** [comment]
@MrLiuYunPing 

Sorry to hear that you are running into issues. To help debug this further, can you provide a code snippet, logs and stacktraces if possible

**Comment by sangios** [comment]
Dear all,

The case that you will see CPU usage too high when turns off wifi, or in airplane mode, or no internet connection.

while (self.runLoopShouldContinue in - (void)openStreams:(id)sender of AWSIoTMQTTClient.m never exits. Also "[runLoopForStreamsThread runMode:NSDefaultRunLoopMode beforeDate:[NSDate distantFuture]];"  without input source will make cpu too high.

**Comment by cbommas** [comment]
@sangios 
I tried to reproduce this issue on my side by running instruments on a iPhone 5s, iOS 9.3.5 and did not see the high CPU usage. Would be great if you could provide details on how you are observing this to help me reproduce it so that I can debug the issue. 

The while loop in the `openStreams` method runs as long as the `runLoopShouldContinue` instance variable is set to `YES`. This run loop handles  MQTT messages and will continue to be executed each time there is a message to be sent, to be received or a timer event to be handled. The `runLoopShouldContinue` variable is set to false in another thread when a disconnect ( see the `disconnect` method in the same class) is issued to the client which causes the while to exit.

My understanding of runloops is that a runloop will exit if there are no input sources/timers scheduled on it  (see https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Multithreading/RunLoopManagement/RunLoopManagement.html#//apple_ref/doc/uid/10000057i-CH16-SW4 ). Can you please provide some more details/share some info on your point about the CPU being too high due to the runloop.

This may be a moot point - in the SDK architecture, this run loop always has multiple input sources and timers scheduled on it.

**Comment by cbommas** [comment]
@sangios 

While I haven't been able to reproduce the high CPU situation, I have made some tweaks to the runloop code as per your reccomendation. I have replaced the `[NSDate distantFuture]` with `[NSDate dateWithTimeIntervalSinceNow:10]`.  So each cycle of the run loop will exit after 10 seconds if no event was received. 

Also, I added an  additional timer to the runloop to make sure that there is always a Timer attached to the runloop. That should prevent it from exiting immediately during a reconnect sequence if (depending on the timing of the threads) there are no input sources or timers attached to the run loop.

These changes will be available in the next rev of the SDK. I will loop back on this thread once the next rev is out.
